File: src/githubclaw/process_manager.py
"""Flat process tree manager for orchestrators and worker agents.

All child processes (orchestrators and workers) are managed as siblings of the
webhook server process.  The manager handles spawning, monitoring exit codes,
idle timeouts, crash detection, concurrency throttling, and the fork PR gate.
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import signal
import time
from dataclasses import dataclass, field
from enum import StrEnum
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from collections.abc import Callable, Coroutine

from githubclaw.constants import (
    DEFAULT_GRACEFUL_DRAIN_SECONDS,
    DEFAULT_MAX_CONCURRENT_AGENTS,
    DEFAULT_PROCESS_TIMEOUT_SECONDS,
    GRACEFUL_DRAIN_POLL_SECONDS,
    MONITOR_CHECK_INTERVAL_SECONDS,
    SIGTERM_GRACE_PERIOD_SECONDS,
)

logger = logging.getLogger(__name__)


class ProcessKind(StrEnum):
    ORCHESTRATOR = "orchestrator"
    WORKER = "worker"


class ProcessState(StrEnum):
    RUNNING = "running"
    FINISHED = "finished"
    CRASHED = "crashed"
    TIMED_OUT = "timed_out"
    KILLED = "killed"


@dataclass
class ManagedProcess:
    """Bookkeeping record for a child process."""

    pid: int
    process: asyncio.subprocess.Process
    kind: ProcessKind
    repo: str
    label: str  # e.g. "coder-repoA-issue42"
    started_at: float = field(default_factory=time.monotonic)
    timeout_seconds: int = DEFAULT_PROCESS_TIMEOUT_SECONDS
    state: ProcessState = ProcessState.RUNNING
    exit_code: int | None = None


@dataclass
class PendingDispatch:
    """A queued dispatch waiting for a free concurrency slot."""

    cmd: str
    args: list[str]
    kind: ProcessKind
    repo: str
    label: str
    timeout_seconds: int


# Read-only agent types exempt from the fork PR gate
_READ_ONLY_AGENT_TYPES = frozenset({"security_reviewer"})


def check_fork_pr_gate(event_payload: dict[str, Any], agent_type: str) -> bool:
    """Return True if the dispatch is allowed, False if it should be blocked.

    Enforcement rules (from SECURITY.md):
      - Not a fork PR -> allow
      - Fork PR with ``githubclaw-approved`` label -> allow
      - Read-only agents (e.g. security_reviewer) -> always allow
      - Otherwise -> block
    """
    pr = event_payload.get("pull_request", {})
    if not pr:
        return True  # Not a PR event at all

    head_repo = pr.get("head", {}).get("repo", {})
    if not head_repo.get("fork", False):
        return True  # Not from a fork

    labels = [label["name"] for label in pr.get("labels", [])]
    if "githubclaw-approved" in labels:
        return True

    return agent_type in _READ_ONLY_AGENT_TYPES


class ProcessManager:
    """Manages all child processes as a flat sibling tree.

    Provides spawn, monitor, idle-timeout, concurrency throttle, graceful
    drain, and force kill capabilities.
    """

    def __init__(
        self,
        max_concurrent_agents: int = DEFAULT_MAX_CONCURRENT_AGENTS,
        on_crash: Callable[[ManagedProcess], Coroutine[Any, Any, None]] | None = None,
        on_timeout: Callable[[ManagedProcess], Coroutine[Any, Any, None]] | None = None,
        on_exit: Callable[[ManagedProcess], Coroutine[Any, Any, None]] | None = None,
    ) -> None:
        self.max_concurrent_agents = max_concurrent_agents
        self._on_crash = on_crash
        self._on_timeout = on_timeout
        self._on_exit = on_exit

        self._processes: dict[int, ManagedProcess] = {}
        self._monitor_task: asyncio.Task[None] | None = None
        self._running = False

        # Queued dispatches waiting for a free slot
        self._pending_dispatches: asyncio.Queue[PendingDispatch] = asyncio.Queue()

    # ------------------------------------------------------------------
    # Properties
    # ------------------------------------------------------------------

    @property
    def active_count(self) -> int:
        return sum(1 for p in self._processes.values() if p.state == ProcessState.RUNNING)

    @property
    def active_processes(self) -> list[ManagedProcess]:
        return [p for p in self._processes.values() if p.state == ProcessState.RUNNING]

    def has_capacity(self) -> bool:
        return self.active_count < self.max_concurrent_agents

    # ------------------------------------------------------------------
    # Spawn
    # ------------------------------------------------------------------

    async def spawn(
        self,
        cmd: str,
        args: list[str],
        kind: ProcessKind,
        repo: str,
        label: str,
        timeout_seconds: int = DEFAULT_PROCESS_TIMEOUT_SECONDS,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
    ) -> ManagedProcess | None:
        """Spawn a child process.

        If the concurrency limit has been reached the dispatch is queued and
        ``None`` is returned.  The queued dispatch will be executed
        automatically when a slot frees up.
        """
        if not self.has_capacity():
            logger.warning(
                "Concurrency limit (%d) reached, queueing dispatch: %s",
                self.max_concurrent_agents,
                label,
            )
            self._pending_dispatches.put_nowait(
                PendingDispatch(
                    cmd=cmd,
                    args=args,
                    kind=kind,
                    repo=repo,
                    label=label,
                    timeout_seconds=timeout_seconds,
                )
            )
            return None

        return await self._do_spawn(cmd, args, kind, repo, label, timeout_seconds, env, cwd)

    async def _do_spawn(
        self,
        cmd: str,
        args: list[str],
        kind: ProcessKind,
        repo: str,
        label: str,
        timeout_seconds: int,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
    ) -> ManagedProcess:
        import os

        merged_env = {**os.environ, **(env or {})}

        proc = await asyncio.create_subprocess_exec(
            cmd,
            *args,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
            env=merged_env,
            cwd=cwd,
        )

        managed = ManagedProcess(
            pid=proc.pid,
            process=proc,
            kind=kind,
            repo=repo,
            label=label,
            timeout_seconds=timeout_seconds,
        )
        self._processes[proc.pid] = managed
        logger.info("Spawned %s process [%s] pid=%d", kind.value, label, proc.pid)
        return managed

    # ------------------------------------------------------------------
    # Monitoring loop
    # ------------------------------------------------------------------

    async def _monitor_loop(self) -> None:
        """Periodically check child processes for exit / timeout."""
        while self._running:
            now = time.monotonic()
            for managed in list(self._processes.values()):
                if managed.state != ProcessState.RUNNING:
                    continue

                # Check if process has exited
                retcode = managed.process.returncode
                if retcode is not None:
                    managed.exit_code = retcode
                    if retcode == 0:
                        managed.state = ProcessState.FINISHED
                        logger.info(
                            "Process [%s] pid=%d finished (exit 0)", managed.label, managed.pid
                        )
                        if self._on_exit:
                            await self._on_exit(managed)
                    else:
                        managed.state = ProcessState.CRASHED
                        logger.error(
                            "Process [%s] pid=%d crashed (exit %d)",
                            managed.label,
                            managed.pid,
                            retcode,
                        )
                        if self._on_crash:
                            await self._on_crash(managed)

                    # Try to drain a pending dispatch now that a slot is free
                    await self._drain_one_pending()
                    continue

                # Check timeout
                elapsed = now - managed.started_at
                if elapsed > managed.timeout_seconds:
                    logger.warning(
                        "Process [%s] pid=%d timed out after %.0fs",
                        managed.label,
                        managed.pid,
                        elapsed,
                    )
                    await self._kill_process(managed)
                    managed.state = ProcessState.TIMED_OUT
                    if self._on_timeout:
                        await self._on_timeout(managed)
                    await self._drain_one_pending()

            await asyncio.sleep(MONITOR_CHECK_INTERVAL_SECONDS)

    async def _drain_one_pending(self) -> None:
        """If there is a queued dispatch and capacity, spawn it."""
        if self._pending_dispatches.empty() or not self.has_capacity():
            return
        try:
            pending = self._pending_dispatches.get_nowait()
        except asyncio.QueueEmpty:
            return
        logger.info("Draining pending dispatch: %s", pending.label)
        await self._do_spawn(
            pending.cmd,
            pending.args,
            pending.kind,
            pending.repo,
            pending.label,
            pending.timeout_seconds,
        )

    # ------------------------------------------------------------------
    # Kill helpers
    # ------------------------------------------------------------------

    @staticmethod
    async def _kill_process(managed: ManagedProcess) -> None:
        """Send SIGTERM then SIGKILL after a grace period."""
        try:
            managed.process.send_signal(signal.SIGTERM)
        except ProcessLookupError:
            return
        try:
            await asyncio.wait_for(managed.process.wait(), timeout=SIGTERM_GRACE_PERIOD_SECONDS)
        except TimeoutError:
            with contextlib.suppress(ProcessLookupError):
                managed.process.kill()
        managed.exit_code = managed.process.returncode

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    def start_monitor(self) -> None:
        """Start the process monitoring background task."""
        if self._monitor_task is not None and not self._monitor_task.done():
            return
        self._running = True
        self._monitor_task = asyncio.get_event_loop().create_task(self._monitor_loop())
        logger.info("Process monitor started")

    async def graceful_drain(self, timeout: float = DEFAULT_GRACEFUL_DRAIN_SECONDS) -> None:
        """Wait for all running processes to finish, up to *timeout* seconds.

        After the timeout, any remaining processes are killed.
        """
        logger.info(
            "Graceful drain: waiting up to %.0fs for %d process(es)", timeout, self.active_count
        )
        deadline = time.monotonic() + timeout

        while self.active_count > 0 and time.monotonic() < deadline:
            await asyncio.sleep(GRACEFUL_DRAIN_POLL_SECONDS)

        # Kill stragglers
        for managed in list(self._processes.values()):
            if managed.state == ProcessState.RUNNING:
                logger.warning("Force-killing straggler [%s] pid=%d", managed.label, managed.pid)
                await self._kill_process(managed)
                managed.state = ProcessState.KILLED

    async def force_kill_all(self) -> None:
        """Immediately kill all running child processes."""
        logger.warning("Force-killing all %d running process(es)", self.active_count)
        for managed in list(self._processes.values()):
            if managed.state == ProcessState.RUNNING:
                await self._kill_process(managed)
                managed.state = ProcessState.KILLED

    async def stop(self) -> None:
        """Stop the monitor loop."""
        self._running = False
        if self._monitor_task is not None:
            self._monitor_task.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await self._monitor_task
            self._monitor_task = None
        logger.info("Process monitor stopped")

    def get_process(self, pid: int) -> ManagedProcess | None:
        return self._processes.get(pid)

    def all_processes(self) -> list[ManagedProcess]:
        return list(self._processes.values())

File: src/process_manager.rs
//! Flat process tree manager for orchestrators and worker agents.
//!
//! All child processes (orchestrators and workers) are managed as siblings of the
//! webhook server process. The manager handles spawning, monitoring exit codes,
//! idle timeouts, crash detection, concurrency throttling, and the fork PR gate.

use std::collections::HashMap;
use std::sync::Arc;
use std::time::Duration;

use tokio::sync::Mutex;
use tokio::time::Instant;

use crate::constants::MONITOR_CHECK_INTERVAL_SECONDS;

/// The kind of child process being managed.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProcessKind {
    Orchestrator,
    Worker,
}

impl std::fmt::Display for ProcessKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            ProcessKind::Orchestrator => write!(f, "orchestrator"),
            ProcessKind::Worker => write!(f, "worker"),
        }
    }
}

/// The lifecycle state of a managed process.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProcessState {
    Running,
    Finished,
    Crashed,
    TimedOut,
    Killed,
}

/// Bookkeeping record for a child process.
pub struct ManagedProcess {
    pub pid: u32,
    pub kind: ProcessKind,
    pub repo: String,
    pub label: String,
    pub started_at: Instant,
    pub timeout_seconds: u64,
    pub state: ProcessState,
    pub exit_code: Option<i32>,
}

/// Read-only agent types exempt from the fork PR gate.
const READ_ONLY_AGENT_TYPES: &[&str] = &["security_reviewer"];

/// Return `true` if the dispatch is allowed, `false` if it should be blocked.
///
/// Enforcement rules (from SECURITY.md):
/// - Not a fork PR -> allow
/// - Fork PR with `githubclaw-approved` label -> allow
/// - Read-only agents (e.g. `security_reviewer`) -> always allow
/// - Otherwise -> block
pub fn check_fork_pr_gate(event_payload: &serde_json::Value, agent_type: &str) -> bool {
    let pr = match event_payload.get("pull_request") {
        Some(pr) if !pr.is_null() => pr,
        _ => return true, // Not a PR event at all
    };

    let is_fork = pr
        .get("head")
        .and_then(|h| h.get("repo"))
        .and_then(|r| r.get("fork"))
        .and_then(|f| f.as_bool())
        .unwrap_or(false);

    if !is_fork {
        return true; // Not from a fork
    }

    // Check for the approved label
    if let Some(labels) = pr.get("labels").and_then(|l| l.as_array()) {
        for label in labels {
            if let Some(name) = label.get("name").and_then(|n| n.as_str()) {
                if name == "githubclaw-approved" {
                    return true;
                }
            }
        }
    }

    // Read-only agents are always allowed on fork PRs
    READ_ONLY_AGENT_TYPES.contains(&agent_type)
}

/// Manages all child processes as a flat sibling tree.
///
/// Provides spawn, monitor, idle-timeout, concurrency throttle, graceful
/// drain, and force kill capabilities.
pub struct ProcessManager {
    pub max_concurrent_agents: usize,
    processes: Arc<Mutex<HashMap<u32, ManagedProcess>>>,
}

impl ProcessManager {
    /// Create a new `ProcessManager` with the given concurrency limit.
    pub fn new(max_concurrent_agents: usize) -> Self {
        Self {
            max_concurrent_agents,
            processes: Arc::new(Mutex::new(HashMap::new())),
        }
    }

    /// Number of processes currently in the `Running` state.
    pub async fn active_count(&self) -> usize {
        let procs = self.processes.lock().await;
        procs
            .values()
            .filter(|p| p.state == ProcessState::Running)
            .count()
    }

    /// Whether there is at least one free concurrency slot.
    pub async fn has_capacity(&self) -> bool {
        self.active_count().await < self.max_concurrent_agents
    }

    /// Return a snapshot of all process PIDs and their states.
    pub async fn all_processes(&self) -> Vec<(u32, ProcessState)> {
        let procs = self.processes.lock().await;
        procs.iter().map(|(&pid, p)| (pid, p.state)).collect()
    }

    /// Register a spawned process for tracking.
    pub async fn register(
        &self,
        pid: u32,
        kind: ProcessKind,
        repo: &str,
        label: &str,
        timeout_seconds: u64,
    ) {
        let managed = ManagedProcess {
            pid,
            kind,
            repo: repo.to_string(),
            label: label.to_string(),
            started_at: Instant::now(),
            timeout_seconds,
            state: ProcessState::Running,
            exit_code: None,
        };
        self.processes.lock().await.insert(pid, managed);
    }

    /// Report that a process has exited with the given exit code.
    pub async fn report_exit(&self, pid: u32, exit_code: i32) {
        if let Some(proc) = self.processes.lock().await.get_mut(&pid) {
            proc.exit_code = Some(exit_code);
            proc.state = if exit_code == 0 {
                ProcessState::Finished
            } else {
                ProcessState::Crashed
            };
        }
    }

    /// Start a background monitor that periodically checks for timed-out processes.
    pub fn start_monitor(self: &Arc<Self>) -> tokio::task::JoinHandle<()> {
        let pm = self.clone();
        tokio::spawn(async move {
            loop {
                tokio::time::sleep(Duration::from_secs(MONITOR_CHECK_INTERVAL_SECONDS)).await;
                let mut procs = pm.processes.lock().await;
                let now = Instant::now();
                for proc in procs.values_mut() {
                    if proc.state == ProcessState::Running {
                        let elapsed = now.duration_since(proc.started_at).as_secs();
                        if elapsed > proc.timeout_seconds {
                            proc.state = ProcessState::TimedOut;
                            tracing::warn!(
                                "Process [{}] pid={} timed out after {}s",
                                proc.label,
                                proc.pid,
                                elapsed
                            );
                        }
                    }
                }
            }
        })
    }
}

// -----------------------------------------------------------------------
// Tests
// -----------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use crate::constants::DEFAULT_MAX_CONCURRENT_AGENTS;
    use serde_json::json;

    // -- check_fork_pr_gate tests --

    #[test]
    fn fork_gate_not_a_pr_event_returns_true() {
        let payload = json!({"action": "opened", "issue": {"number": 1}});
        assert!(check_fork_pr_gate(&payload, "coder"));
    }

    #[test]
    fn fork_gate_non_fork_pr_returns_true() {
        let payload = json!({
            "pull_request": {
                "head": {"repo": {"fork": false}},
                "labels": []
            }
        });
        assert!(check_fork_pr_gate(&payload, "coder"));
    }

    #[test]
    fn fork_gate_fork_pr_with_approved_label_returns_true() {
        let payload = json!({
            "pull_request": {
                "head": {"repo": {"fork": true}},
                "labels": [{"name": "githubclaw-approved"}]
            }
        });
        assert!(check_fork_pr_gate(&payload, "coder"));
    }

    #[test]
    fn fork_gate_fork_pr_without_label_blocks() {
        let payload = json!({
            "pull_request": {
                "head": {"repo": {"fork": true}},
                "labels": []
            }
        });
        assert!(!check_fork_pr_gate(&payload, "coder"));
    }

    #[test]
    fn fork_gate_fork_pr_security_reviewer_always_allowed() {
        let payload = json!({
            "pull_request": {
                "head": {"repo": {"fork": true}},
                "labels": []
            }
        });
        assert!(check_fork_pr_gate(&payload, "security_reviewer"));
    }

    #[test]
    fn fork_gate_fork_pr_coder_blocked() {
        let payload = json!({
            "pull_request": {
                "head": {"repo": {"fork": true}},
                "labels": [{"name": "some-other-label"}]
            }
        });
        assert!(!check_fork_pr_gate(&payload, "coder"));
    }

    // -- ProcessManager tests --

    #[test]
    fn process_manager_new_with_capacity() {
        let pm = ProcessManager::new(4);
        assert_eq!(pm.max_concurrent_agents, 4);
    }

    #[tokio::test]
    async fn active_count_starts_at_zero() {
        let pm = ProcessManager::new(DEFAULT_MAX_CONCURRENT_AGENTS);
        assert_eq!(pm.active_count().await, 0);
    }

    #[tokio::test]
    async fn has_capacity_returns_true_when_empty() {
        let pm = ProcessManager::new(DEFAULT_MAX_CONCURRENT_AGENTS);
        assert!(pm.has_capacity().await);
    }

    #[tokio::test]
    async fn register_adds_process_to_tracking() {
        let pm = ProcessManager::new(4);
        assert_eq!(pm.active_count().await, 0);

        pm.register(1001, ProcessKind::Worker, "owner/repo", "test-worker", 3600)
            .await;
        assert_eq!(pm.active_count().await, 1);

        let procs = pm.all_processes().await;
        assert_eq!(procs.len(), 1);
        assert_eq!(procs[0], (1001, ProcessState::Running));
    }

    #[tokio::test]
    async fn report_exit_updates_state_success() {
        let pm = ProcessManager::new(4);
        pm.register(2001, ProcessKind::Orchestrator, "owner/repo", "orch", 3600)
            .await;

        pm.report_exit(2001, 0).await;
        let procs = pm.all_processes().await;
        assert_eq!(procs[0].1, ProcessState::Finished);
        // No longer counted as active
        assert_eq!(pm.active_count().await, 0);
    }

    #[tokio::test]
    async fn report_exit_updates_state_crash() {
        let pm = ProcessManager::new(4);
        pm.register(3001, ProcessKind::Worker, "owner/repo", "worker", 3600)
            .await;

        pm.report_exit(3001, 1).await;
        let procs = pm.all_processes().await;
        assert_eq!(procs[0].1, ProcessState::Crashed);
    }

    #[tokio::test]
    async fn has_capacity_respects_registered_processes() {
        let pm = ProcessManager::new(2);
        assert!(pm.has_capacity().await);

        pm.register(100, ProcessKind::Worker, "r", "w1", 3600).await;
        assert!(pm.has_capacity().await); // 1 of 2

        pm.register(101, ProcessKind::Worker, "r", "w2", 3600).await;
        assert!(!pm.has_capacity().await); // 2 of 2 — full

        // Finishing one frees capacity
        pm.report_exit(100, 0).await;
        assert!(pm.has_capacity().await); // 1 of 2 again
    }
}
